docs: Update Indexer to Conduit migration guide. (#26)
algorand/conduit: docs/tutorials/IndexerMigration.md

We've created a new API on algod, `GET v2/deltas/{round}`, which exposes per-round deltas. It is a direct serialization
of the data which is stored in the cache in algod, and it is backward compatible with the postgresql serialization that
the Indexer database currently uses.

Now instead of running a local ledger, all that is required to run an Indexer pipeline via Conduit is a Follower algod
node. Conduit's algod importer plugin will coordinate usage of both the sync round APIs and the deltas API in order to
fetch all required data directly from the algod node, and ensure that the node is kept in sync with the pipeline round.

## Altering Architecture of Indexer Applications

Let's take a look at a theoretical application more closely and see what in particular we need to do to alter it to use
Conduit. 

Here's a diagram of a containerized application (leaving out some details such as load balancers, and scaled out API
nodes).

```mermaid
graph LR;
    algod["Archival Alogd"]
    indexer["Indexer"]
    psql["Postgresql"]
    subgraph "Persistent Volume"
        ll["Local Ledger"];
    end
    
    indexer-->algod;
    indexer-->psql;
    indexer-->ll;
```

> ### Step 1: Convert from Archival to Follower Node

Reviewer: ```suggestion
### Step 1: Setup a Follower Node
```

In our hypothetical existing architecture, we likely have a persistent volume storing the algod data directory for our
Archival node. It is probably still a good idea to maintain the algod's data in a persistent volume, though the amount
of storage required will be minimal in comparison.

If you have an existing Archival node which is on the same round as your Indexer, you can simply stop algod, alter your

> config as follows, and restart it.

Reviewer: Did you try this? What happens when you increase `MaxAcctLookback` from the default?

Author: I did try it but I didn't pay attention to what happens to the cache on startup. I was mostly paying attention to whether or not older rounds than the last 1k would be deleted from the ledger on startup (and they are).

```yaml
Archival: false,
MaxAcctLookback: 256,

> CatchupParallelBlocks: 256,

Reviewer: I've been leaving this at 32 with good results.
```suggestion
CatchupParallelBlocks: 32,
```

EnableFollowMode: true
```

> Your node will then drop all but the last 1000 rounds from your database, and when will have `2 * MaxAcctLookback`

Reviewer 2: ```suggestion
Your node will then drop all but the last 1000 rounds from your database, and will have `2 * MaxAcctLookback`
```

Follower nodes provide the same methods of catchup as regular algod nodes, but will have the sync round set to the most
recent ledger round upon startup.

In order to manually catchup your node from 0 to 25 million, for example, you will need to call the sync round API to
update the ledger constraint:
```
curl -X POST -H "X-Algo-API-TOKEN:$ALGOD_TOKEN" http://$ALGOD_ADDR/v2/ledger/sync/25000000
```

Now you can run fast catchup on your node to the closest catchpoint prior to the desired sync round.
```
goal node -d $ALGOD_DIR catchup 25000000#EOX5UYQV4IXTGYQCIDR7ZLUK6WZGDC5EG6PYQGBG6PBYNAQUPN2Q
```
Once catchup is complete we will resume normal catchup. If your catchpoint is not exactly the round your database is on,
you will need to wait for normal catchup to get your node to the desired point in time. The sync round we set earlier will ensure that the node does not
advance past the round Conduit needs before we're ready. 

### Step 2: Remove the Local Ledger
Because our Conduit pipeline will use the Follower node's state delta API, we no longer need our local ledger persistent 

> volume. Simply remove it.

Reviewer: ```suggestion
volume. Simply remove it by deleting your indexer data directory. **If you're using a configuration YAML file be sure to back that up before deleting the data directory.**
```

Reviewer 3: +1
also suggest removing "Simply"

### Step 3: Refactor our Indexer Writer to Conduit
You're free to capture any data you like using Conduit. I'd encourage you to take a look at the filter processor and see
if you can reduce the amount of data you store in your database by removing non-relevant data.

> In order to have parity with the previous Indexer, which stored all data, you can use the following Conduit configuration.

Reviewer: What do you think about suggesting `./cmd/conduit/conduit init --importer algod --exporter postgresql --data new_data_dir` instead of copy/pasting the config?

> In order to have parity with the previous Indexer, which stored all data, you can use the following Conduit configuration.

Reviewer 3: "If you would like to maintain parity with the legacy Indexer which stored all data,  you can ..."

```yaml
retry-count: 10
retry-delay: "1s"

> data-dir: $DATA_DIR

Reviewer 2: ```suggestion
data-dir: $CONDUIT_DATA_DIR
```
to be consistent with  `conduit -d $CONDUIT_DATA_DIR`

next-round-override: $DB_ROUND
hide-banner: true
metrics:
  prefix: "conduit",

> log-level: logger.GetLevel().String(),

Reviewer: ```suggestion
log-level: "INFO"
```

Conduit. 

Here's a diagram of a containerized application (leaving out some details such as load balancers, and scaled out API
nodes).

```mermaid
graph LR;
    algod["Archival Alogd"]
    indexer["Indexer"]
    psql["Postgresql"]
    subgraph "Persistent Volume"
        ll["Local Ledger"];
    end
    
    indexer-->algod;
    indexer-->psql;
    indexer-->ll;
```

### Step 1: Convert from Archival to Follower Node
In our hypothetical existing architecture, we likely have a persistent volume storing the algod data directory for our
Archival node. It is probably still a good idea to maintain the algod's data in a persistent volume, though the amount
of storage required will be minimal in comparison.

If you have an existing Archival node which is on the same round as your Indexer, you can simply stop algod, alter your
config as follows, and restart it.
```yaml
Archival: false,
MaxAcctLookback: 256,
CatchupParallelBlocks: 32,
EnableFollowMode: true
```
Your node will then drop all but the last 1000 rounds from your database, and will have `2 * MaxAcctLookback`
rounds in its cache.

> However, if your algod node's round is too far beyond your Indexer's you will need to catchup your node from scratch.

Reviewer 3: how is a user supposed to detect this?

Reviewer: There's an error message in the log
> ledger state delta not found: node round (%d), required round (%d): verify follower node configuration and ensure follower node has its sync round set to the required round, re-deploying the follower node may be necessary

Author: Also updated with info below on how to use `goal` for algod and `/health` for indexer to determine which round they are on.

Follower nodes provide the same methods of catchup as regular algod nodes, but will have the sync round set to the most
recent ledger round upon startup.

In order to manually catchup your node from 0 to 25 million, for example, you will need to call the sync round API to
update the ledger constraint:
```
curl -X POST -H "X-Algo-API-TOKEN:$ALGOD_TOKEN" http://$ALGOD_ADDR/v2/ledger/sync/25000000
```

Now you can run fast catchup on your node to the closest catchpoint prior to the desired sync round. For a list of catchpoints,
you can reference the following:
* [Mainnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/mainnet_catchpoints.txt)
* [Testnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/testnet_catchpoints.txt)
* [Betanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/betanet_catchpoints.txt)

```
goal node -d $ALGOD_DIR catchup 25000000#EOX5UYQV4IXTGYQCIDR7ZLUK6WZGDC5EG6PYQGBG6PBYNAQUPN2Q
```

> Once catchup is complete we will resume normal catchup. If your catchpoint is not exactly the round your database is on,

Reviewer 3: ```suggestion
Once catchup is complete the node will automatically resume normal catchup. If your catchpoint is not exactly the round your database is on,
```

(propose new blocks), or broadcast transactions to be included in new blocks. It will only, as named, follow the chain.

Follower mode algod provides 2 key features required by Conduit.
### Sync Rounds
In order to remove the requirement of algod being Archival, we needed a way to ensure that algod would have the data
available for the particular round that our Conduit pipeline was importing at any given time. We've done this through
the concept of a `sync` round. 

The sync round is a parameter which Conduit provides to algod. It specifies the particular round that we want algod to
ensure is kept in the cache until we've finished running it through our Conduit pipeline. When a sync round is set on a
Follower node, that node will no-op network fetches of new blocks for any block which would cause data on the sync round
to fall out of the cache.

Sync rounds operate via a set of APIs on algod.

#### `GET    /v2/ledger/sync`
Retrieves the sync round. This call will return a `404 Not Found` if the sync round is not currently set.
#### `DELETE /v2/ledger/sync`
Removes the sync round constraint. If no sync round is set this will no-op.

> #### `POST   /v2/ledger/sync/{round}`

Reviewer: Good point about the official docs. What do you think about linking to that directly instead of documenting it here?

### Step 2: Remove the Local Ledger
Because our Conduit pipeline will use the Follower node's state delta API, we no longer need our local ledger persistent 
volume. It can be removed.

### Step 3: Refactor our Indexer Writer to Conduit
You're free to capture any data you like using Conduit. I'd encourage you to take a look at the [filter processor](./FilterDeepDive.md) and see
if you can reduce the amount of data you store in your database by removing non-relevant data.

If you would like to maintain parity with the legacy Indexer, which stored all data, you can use the `conduit init` command to create
an initial Conduit config.
```
conduit init --importer algod --exporter postgresql --data new_data_dir
```
Here is an example of a basic config which runs the `algod` importer and `postgresql` exporter.
```yaml
retry-count: 10
retry-delay: "1s"
data-dir: $CONDUIT_DATA_DIR
next-round-override: $DB_ROUND
hide-banner: true
metrics:
  prefix: "conduit",
log-level: "INFO"
importer:
  name: "algod",
  config: 
    "netaddr": $ALGOD_ADDR,
    "token": $ALGOD_TOKEN,
    "mode": "follower",
exporter:
  name: "postgresql",
  config:
    "connection-string": $PGSQL_CONNECTION_STRING,
```

> Then run Conduit, `conduit -d $CONDUIT_DATA_DIR`!

Reviewer: Do they need to set `--next-round-override` the first time?
